routetable/route_table.go

```diff
@@ -304,226 +448,345 @@ func (r *RouteTable) Apply() error {
+			"Leaving it dirty, requiring a full sync.")
+		r.markIfaceForUpdate(ifaceName, true)
+	}
 
 	r.cleanUpPendingConntrackDeletions()
 
 	// Don't return a failure if there are only interfaces in the cleanup grace period.
 	// They'll be retried on the next invocation (the route refresh timer), and we mustn't
 	// count them as Sync Errors.
-	if r.dirtyIfaces.Len() > graceIfaces {
+	if len(r.ifaceNameToUpdateType) > graceIfaces {
 		r.logCxt.Warn("Some interfaces still out-of sync.")
-		r.inSync = false
 		return UpdateFailed
 	}
 
 	return nil
 }
 
-func (r *RouteTable) ensureL2Dataplane(linkAttrs *netlink.LinkAttrs, target L2Target) error {
-	// For each L2 entry we need to program, program it.
-	// Add a static ARP entry.
-	a := &netlink.Neigh{
-		LinkIndex:    linkAttrs.Index,
-		State:        netlink.NUD_PERMANENT,
-		Type:         syscall.RTN_UNICAST,
-		IP:           target.GW.AsNetIP(),
-		HardwareAddr: target.VTEPMAC,
+func (r *RouteTable) syncRoutesForLink(ifaceName string, fullSync bool) error {
+	startTime := time.Now()
+	defer func() {
+		perIfaceSyncTime.Observe(r.time.Since(startTime).Seconds())
+	}()
+	logCxt := r.logCxt.WithField("ifaceName", ifaceName)
+	logCxt.Debug("Syncing interface routes")
+
+	// If necessary perform a full resync. This will return a set of routes that need deleting and will update the
+	// deltas to fix any discrepancies with the expected configuration. If this errors, we still apply the deltas
+	// first because this allows us to tidy up configuration for interfaces that no longer have any routes associated
+	// with them.
+	var routesToDelete []netlink.Route
+	var resyncErr error
+	if fullSync {
+		// Performing a full re-sync.  Start by applying the deltas so that we don't delete routes that are required.
+		logCxt.Debug("Reconcile against kernel programming")
+		_, _ = r.applyRouteDeltas(ifaceName)
+
+		// Now do the resync - this will update our deltas again based on what is not programmed (it's a little bit
+		// circuitous, but simplifies the code paths for resync and delta processing).
+		if routesToDelete, resyncErr = r.fullResyncRoutesForLink(logCxt, ifaceName); resyncErr != nil && resyncErr != IfaceGrace {
+			// If we hit anything other than an interface-in-grace error, exit now.
+			r.logCxt.WithError(resyncErr).Info("Hit error doing kernel reconciliation")
+			return r.filterErrorByIfaceState(ifaceName, resyncErr, UpdateFailed)
+		}
 	}
-	if err := netlink.NeighSet(a); err != nil {
+
+	// Any deleted route that is not being replaced by a route with the same CIDR should have the corresponding
+	// conntrack entry removed.
+	defer func() {
+		cidrsToTarget := r.ifaceNameToTargets[ifaceName]
+		for _, route := range routesToDelete {
+			if cidr := ip.CIDRFromIPNet(route.Dst); cidr == nil {
+				// No parseable CIDR destination in route.
+			} else if _, ok := cidrsToTarget[cidr]; !ok {
+				// Route is deleted and CIDR should not be routable anymore - remove conntrack entries.
```

Comment:
Not sure if this condition is right.  The point of deleting conntrack entries is in case an IP gets reused:
* workload A has 10.0.0.1
* felix gets disconnected form datastore/whatever for 30s
* workload A is removed
* workload B gets created with same ip 10.0.0.1
* felix wakes up, goes through route reconciliation: we _want_ to remove conntrack entries for workload A's use of 10.0.0.1 so that workload B can't re-use them

Reply:
@fasaxc :  Doesn't this check just say "only kick off contract deletion if the CIDR is no longer valid for the interface" - which I think is ok no?  In the case of IP reuse the CIDR will not be valid for the interface anymore.

Reply:
Agreed, I'd misunderstood